### PR TITLE
Fix 02050_client_profile_events flakiness

### PR DESCRIPTION
**Pull request by azat**
Changelog:
- Do not loose ProfileEvents in case of thread destroyed before
- Merge ProfileEvents in case they were not printed

CI: https://s3.amazonaws.com/clickhouse-test-reports/32303/24751e7d45d94541be854c86ce46d65c2e0f66da/stateless_tests__thread__actions_.html

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Cc: @novikd

### CODE REVIEW

**Comment by azat** [comment]
- [BuilderDebUBsan](https://github.com/ClickHouse/ClickHouse/runs/4513144541?check_suite_focus=true)

Some internal github actions issue:

```
  Error: fatal: Unable to create '/home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/.git/modules/contrib/aws/index.lock': File exists.
...
  Unable to checkout '00b03604543367d7e310cb0993973fdcb723ea79' in submodule path 'contrib/aws'
```

Will do a rebase.

File: src/Client/ClientBase.cpp [hunk 1]
```diff
@@ -105,6 +108,100 @@ namespace ProfileEvents
...
+            dst_column_name.getDataAt(dst_row),
+            dst_column_host_name.getDataAt(dst_row),
+            dst_array_thread_id[dst_row],
+        };
+
+        if (auto it = rows_by_name.find(id); it != rows_by_name.end())
+        {
+            size_t src_row = it->second;
+            dst_array_current_time[dst_row] = src_array_current_time[src_row];
+            dst_array_value[dst_row] = src_array_value[src_row];
```
Comment (novikd) [inline src/Client/ClientBase.cpp:177]: Looks like a mistake, value is updated in the switch statement.

**Review by novikd — APPROVED** [review]

File: src/Client/ClientBase.cpp (hunk 1 above)
Comment (azat) [reply src/Client/ClientBase.cpp]: Indeed, nice catch, fixed.

**Comment by azat** [comment]
>Integration tests (asan, actions) [2/3] — fail: 1, passed: 543, flaky: 0

- `test_storage_kerberized_kafka/test.py::test_kafka_json_as_string_no_kdc` - #32644

>Integration tests (release, actions) [1/2] — fail: 1, passed: 746, flaky: 0

- `test_storage_s3/test.py::test_seekable_formats` - #32825

>[Integration tests (thread, actions) [1/4] — fail: 101, passed: 1, flaky: 0](https://s3.amazonaws.com/clickhouse-test-reports/32493/3da2e1daa0b143096f85c1b3cf92999428981be9/integration_tests__thread__actions__[1/4].html)

```
compose.cli.verbose_proxy.proxy_callable: docker pull <- ('clickhouse/integration-test', tag='latest', stream=True, platform=None)
compose.cli.errors.log_api_error: Get "https://registry-1.docker.io/v2/": net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```

- #32772 

> [Integration tests (thread, actions) [2/4] — fail: 1, passed: 459, flaky: 2](https://s3.amazonaws.com/clickhouse-test-reports/32493/3da2e1daa0b143096f85c1b3cf92999428981be9/integration_tests__thread__actions__[2/4].html)

- `test_replicated_fetches_timeouts/test.py::test_no_stall`

> Stress test (debug, actions) — Killed by signal (in clickhouse-server.log)

- `Chunk should have AggregatedChunkInfo in GroupingAggregatedTransform` - #30692

> Stress test (thread, actions) — Killed by signal (in clickhouse-server.log)

```
==516==ERROR: ThreadSanitizer failed to allocate 0x12000 (73728) bytes of ThreadSignalContext (error code: 12)
ERROR: Failed to mmap
```

> Stress test (undefined, actions) — check_status.tsv doesn't exists

Some internal issue?

But let's run CI one more time (just in case).

**Comment by azat** [comment]
Now better:

>AST fuzzer (debug, actions) — Task failed: $?=210

Hm, some issue in `getaddrinfo`

<details>

```
2021.12.16 08:45:41.274358 [ 230 ] {} <Trace> BaseDaemon: Received signal 5
2021.12.16 08:45:41.274778 [ 234 ] {} <Fatal> BaseDaemon: ########################################
2021.12.16 08:45:41.274872 [ 234 ] {} <Fatal> BaseDaemon: (version 21.13.1.538, build id: 2D910F2D2502C654) (from thread 99) (no query) Received signal Trace/breakpoint trap (5)
2021.12.16 08:45:41.274933 [ 234 ] {} <Fatal> BaseDaemon: 
2021.12.16 08:45:41.275024 [ 234 ] {} <Fatal> BaseDaemon: Stack trace: 0x7ff1a47f108f 0x7ff1a46f58b8 0x7ff1a47f05fa 0x7ff1a46f4861 0x7ff1a46f58b8 0x7ff1a46f5983 0x7ff1a46f4995 0x7ff1a46d862c 0x7ff1a46d8ed9 0x7ff1a46994f5 0x7ff1a469b0d9 0x27d23d34 0x27d24f16 0x27cf0203 0x27cf0158 0x157e1982 0x27d79645 0x157d8956 0x27d9a2d3 0x157d67e7 0x156f4e65 0x7ff1a45b90b3 0x156f4a6e
2021.12.16 08:45:41.275106 [ 234 ] {} <Fatal> BaseDaemon: 4. ? @ 0x7ff1a47f108f in ?
2021.12.16 08:45:41.275153 [ 234 ] {} <Fatal> BaseDaemon: 5. _dl_catch_exception @ 0x7ff1a46f58b8 in ?
2021.12.16 08:45:41.275189 [ 234 ] {} <Fatal> BaseDaemon: 6. ? @ 0x7ff1a47f05fa in ?
2021.12.16 08:45:41.275226 [ 234 ] {} <Fatal> BaseDaemon: 7. ? @ 0x7ff1a46f4861 in ?
2021.12.16 08:45:41.275264 [ 234 ] {} <Fatal> BaseDaemon: 8. _dl_catch_exception @ 0x7ff1a46f58b8 in ?
2021.12.16 08:45:41.275302 [ 234 ] {} <Fatal> BaseDaemon: 9. _dl_catch_error @ 0x7ff1a46f5983 in ?
2021.12.16 08:45:41.275340 [ 234 ] {} <Fatal> BaseDaemon: 10. __libc_dlopen_mode @ 0x7ff1a46f4995 in ?
2021.12.16 08:45:41.275395 [ 234 ] {} <Fatal> BaseDaemon: 11. ? @ 0x7ff1a46d862c in ?
2021.12.16 08:45:41.275433 [ 234 ] {} <Fatal> BaseDaemon: 12. __nss_lookup_function @ 0x7ff1a46d8ed9 in ?
2021.12.16 08:45:41.275470 [ 234 ] {} <Fatal> BaseDaemon: 13. ? @ 0x7ff1a46994f5 in ?
2021.12.16 08:45:41.275507 [ 234 ] {} <Fatal> BaseDaemon: 14. getaddrinfo @ 0x7ff1a469b0d9 in ?
2021.12.16 08:45:41.362055 [ 234 ] {} <Fatal> BaseDaemon: 15. ./obj-x86_64-linux-gnu/../contrib/poco/Net/src/DNS.cpp:79: Poco::Net::DNS::hostByName(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int) @ 0x27d23d34 in /workspace/clickhouse
2021.12.16 08:45:41.445293 [ 234 ] {} <Fatal> BaseDaemon: 16. ./obj-x86_64-linux-gnu/../contrib/poco/Net/src/DNS.cpp:194: Poco::Net::DNS::thisHost() @ 0x27d24f16 in /workspace/clickhouse
2021.12.16 08:45:41.517369 [ 234 ] {} <Fatal> BaseDaemon: 17. ./obj-x86_64-linux-gnu/../base/base/getFQDNOrHostName.cpp:11: (anonymous namespace)::getFQDNOrHostNameImpl() @ 0x27cf0203 in /workspace/clickhouse
2021.12.16 08:45:41.578008 [ 234 ] {} <Fatal> BaseDaemon: 18. ./obj-x86_64-linux-gnu/../base/base/getFQDNOrHostName.cpp:23: getFQDNOrHostName() @ 0x27cf0158 in /workspace/clickhouse
2021.12.16 08:45:41.670984 [ 234 ] {} <Fatal> BaseDaemon: 19. ./obj-x86_64-linux-gnu/../programs/server/Server.cpp:769: DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) @ 0x157e1982 in /workspace/clickhouse
2021.12.16 08:45:41.763793 [ 234 ] {} <Fatal> BaseDaemon: 20. ./obj-x86_64-linux-gnu/../contrib/poco/Util/src/Application.cpp:334: Poco::Util::Application::run() @ 0x27d79645 in /workspace/clickhouse
2021.12.16 08:45:41.900908 [ 234 ] {} <Fatal> BaseDaemon: 21. ./obj-x86_64-linux-gnu/../programs/server/Server.cpp:401: DB::Server::run() @ 0x157d8956 in /workspace/clickhouse
2021.12.16 08:45:41.978903 [ 234 ] {} <Fatal> BaseDaemon: 22. ./obj-x86_64-linux-gnu/../contrib/poco/Util/src/ServerApplication.cpp:611: Poco::Util::ServerApplication::run(int, char**) @ 0x27d9a2d3 in /workspace/clickhouse
2021.12.16 08:45:42.115231 [ 234 ] {} <Fatal> BaseDaemon: 23. ./obj-x86_64-linux-gnu/../programs/server/Server.cpp:179: mainEntryClickHouseServer(int, char**) @ 0x157d67e7 in /workspace/clickhouse
2021.12.16 08:45:42.132032 [ 234 ] {} <Fatal> BaseDaemon: 24. ./obj-x86_64-linux-gnu/../programs/main.cpp:378: main @ 0x156f4e65 in /workspace/clickhouse
2021.12.16 08:45:42.132085 [ 234 ] {} <Fatal> BaseDaemon: 25. __libc_start_main @ 0x7ff1a45b90b3 in ?
2021.12.16 08:45:46.044067 [ 234 ] {} <Fatal> BaseDaemon: 26. _start @ 0x156f4a6e in /workspace/clickhouse
2021.12.16 08:45:47.156760 [ 234 ] {} <Fatal> BaseDaemon: Calculated checksum of the binary: 27B40AB4F18BD9CC69E2165982CA0BF4. There is no information about the reference checksum.
2021.12.16 08:45:47.156830 [ 234 ] {} <Information> SentryWriter: Not sending crash report
```

</details>
>Integration tests (asan, actions) [2/3] — fail: 1, passed: 546, flaky: 0

- test_storage_kerberized_kafka/test.py::test_kafka_json_as_string_no_kdc - #kafak

>Integration tests (asan, actions) [3/3] — fail: 1, passed: 626, flaky: 1

- test_s3_zero_copy_replication/test.py::test_s3_zero_copy_replication[s3]

>Integration tests (release, actions) [2/2] — fail: 1, passed: 929, flaky: 1

- test_merge_tree_blob_storage/test.py::test_freeze_unfreeze

```
E           helpers.client.QueryRuntimeException: Client failed! Return code: 233, stderr: Received exception from server (version 21.13.1):
E           Code: 1001. DB::Exception: Received from 172.16.12.3:9000. DB::Exception: Azure::Storage::StorageException: The specified blob does not exist.
```

**Comment by alexey-milovidov** [comment]
@azat This pull request slows down short queries.

**Pull request merged**